kpi/serializers/v2/asset_file.py

```diff
@@ -40,8 +46,262 @@ class Meta:
             'user',
             'user__username',
             'file_type',
-            'name',
+            'description',
             'date_created',
             'content',
             'metadata',
         )
+
+    def get_url(self, obj):
+        return reverse('asset-file-detail',
+                       args=(obj.asset.uid, obj.uid),
+                       request=self.context.get('request', None))
+
+    def get_content(self, obj, *args, **kwargs):
+        return reverse('asset-file-content',
+                       args=(obj.asset.uid, obj.uid),
+                       request=self.context.get('request', None))
+
+    def to_internal_value(self, data):
+        """
+        Overrides parent method to add base64 encoded string to validated data
+        if it exists.
+        """
+        ret = super().to_internal_value(data)
+        # `base64Encoded` is not a valid field, thus it is discarded by Django
+        # validation process. We add it here to be able to access it in
+        # `.validate()` and use our custom validation.
+        try:
+            ret['base64Encoded'] = data['base64Encoded']
+        except KeyError:
+            pass
+        return ret
+
+    def validate(self, attr):
+        self.__file_type = attr['file_type']  # noqa
+
+        metadata = self._get_metadata(attr.get('metadata'))
+        validated_field = self._validate_media_content_method(attr, metadata)
+        # Call the validator related to `validated_field`, either:
+        # - `self._validate_content()`
+        # - `self._validate_base64Encoded()`
+        # - `self._validate_redirect_url()`
+        validator = getattr(self, f'_validate_{validated_field}')
+        validator(attr, metadata)
+
+        # Common validators
+        filename = metadata['filename']
+        self.__validate_mime_type(filename, validated_field)
+        self.__validate_extension(filename, validated_field)
+        self._validate_duplicate(filename, validated_field)
+
+        # Remove `'base64Encoded'` from attributes passed to the model
+        attr.pop('base64Encoded', None)
+
+        return attr
+
+    def _get_metadata(self, metadata: Union[str, dict]) -> dict:
+        """
+        `metadata` parameter can be sent as a stringified JSON or in pure JSON.
```

Comment:
Done

Reply:
For the record, tests use stringified JSON too. Because requests are sent as multipart body. 

> AssertionError: Test data contained a dictionary value for key 'metadata', but multipart uploads do not support nested data. You may want to consider using format='json' in this test case.

Reply:
Huh, good to know. Maybe something we should change in the future.